base/distributed/cluster.jl

```diff
@@ -210,12 +210,46 @@ end
 # The master process uses this to connect to the worker and subsequently
 # setup a all-to-all network.
 function read_worker_host_port(io::IO)
-    while true
-        conninfo = readline(io)
+    t0 = time()
+
+    # Wait at most for JULIA_WORKER_TIMEOUT seconds to read host:port
+    # info from the worker
+    timeout = worker_timeout()
+
+    # We expect the first line to contain the host:port string. However, as
+    # the worker may be launched via ssh or a cluster manager like SLURM,
+    # ignore any informational / warning lines printed by the launch command.
+    # If we do not find the host:port string in the first 1000 lines, treat it
+    # as an error.
+
+    ntries = 1000
+    while ntries > 0
+        readtask = @schedule readline(io)
+        yield()
+        while !istaskdone(readtask) && ((time() - t0) < timeout)
+            sleep(0.05)
+        end
+        !istaskdone(readtask) && break
+
+        conninfo = wait(readtask)
+        if conninfo == "" && !isopen(io)
+            error("Unable to read host:port string from worker. Launch command exited with error?")
```

Comment:
In all probability the launch command exited with an error. However we cannot be sure that it actually exited with a non-zero code.